src/DssSpell.t.sol

```diff
@@ -646,40 +628,97 @@ contract DssSpellTest is DSTest, DSMath {
+   }
+
+   function testFailTooEarly() public {
+       vote();
+       scheduleWaitAndCastFailEarly();
+   }
+
+   function testFailTooLate() public {
+       vote();
+       scheduleWaitAndCastFailLate();
+   }
+
+   function testOnTime() public {
+       vote();
+       scheduleWaitAndCast();
+   }
+
+    // function testSpellIsCast() public {
+    //     string memory description = new DssSpell().description();
+    //     assertTrue(bytes(description).length > 0);
+    //     // DS-Test can't handle strings directly, so cast to a bytes32.
+    //     assertEq(stringToBytes32(spell.description()),
+    //             stringToBytes32(description));
+
+    //     if(address(spell) != address(MAINNET_SPELL)) {
+    //         assertEq(spell.expiration(), (now + 4 days + 2 hours));
+    //     } else {
+    //         assertEq(spell.expiration(), (SPELL_CREATED + 4 days + 2 hours));
+    //     }
+
+    //     vote();
+    //     scheduleWaitAndCast();
+    //     assertTrue(spell.done());
+
+    //     checkSystemValues(afterSpell);
+
+    //     checkCollateralValues(afterSpell);
+    // }
+
+    function test_nextCastTime() public {
+        hevm.warp(1606161600); // Nov 23, 3:00pm (could be cast Nov 26)
 
         vote();
-        scheduleWaitAndCast();
-        assertTrue(spell.done());
+        spell.schedule();
+
+        uint256 monday_9am  = 1606744800; // Nov 30
+
+        /*** Day tests */
+        hevm.warp(monday_9am); // Monday, 9am
+        assertEq(spell.nextCastTime(), monday_9am);
+
+        hevm.warp(monday_9am - 1 days); // Sunday, 9am
+        assertEq(spell.nextCastTime(), monday_9am);
+
+        hevm.warp(monday_9am - 2 days); // Saturday, 9am
+        assertEq(spell.nextCastTime(), monday_9am);
+
+        hevm.warp(monday_9am - 3 days); // Friday, 9am
+        assertEq(spell.nextCastTime(), monday_9am - 3 days); // Able to cast
+
+        /*** Time tests ***/
+        uint256 castTime;
+
+        for(uint i = 0; i < 5; i++) {
+            castTime = monday_9am + i * 1 days; // Next day at 9am
+            hevm.warp(castTime - 1 seconds); // 8:59am
+            assertEq(spell.nextCastTime(), castTime);
+
+            hevm.warp(castTime + 7 hours + 1 seconds); // 4:01pm
+            if (i < 4) {
+                assertEq(spell.nextCastTime(), monday_9am + (i + 1) * 1 days); // Next day at 9am
+            } else {
+                assertEq(spell.nextCastTime(), monday_9am + 7 days); // Next monday at 9am (friday case)
+            }
+        }
+    }
+
+    function testFail_notScheduled() public {
+        spell.nextCastTime();
+    }
 
-        checkSystemValues(afterSpell);
+    function test_use_eta() public {
+        hevm.warp(1606161600); // Nov 23, 3pm (could be cast Nov 26)
 
-        checkCollateralValues(afterSpell);
+        vote();
+        spell.schedule();
+
+        uint castTime = spell.nextCastTime();
+        assertEq(castTime, spell.eta()); // Nov 26, 3pm
     }
-}
+}
```

Comment:
There should be a newline here. You might want to update your editor to automatically add a newline at the end of a file.

Reply:
Added